contracts/modules/LoanClosingsShared.sol

```diff
@@ -319,28 +321,32 @@ contract LoanClosingsShared is
      * @param returnTokenIsCollateral Defines if the remainder should be paid
      *   out in collateral tokens or underlying loan tokens.
      *
-     * @return loanCloseAmount The amount of the collateral token of the loan.
-     * @return withdrawAmount The withdraw amount in the collateral token.
-     * @return withdrawToken The loan token address.
+     * @return closeWithSwapReturn struct which consist of:
+     * - uint256 loanCloseAmount -- The amount of the collateral token of the loan.
+     * - uint256 withdrawAmount -- The withdraw amount in the collateral token.
+     * - address withdrawToken -- The loan token address.
      * */
     function _closeWithSwap(
         bytes32 loanId,
         address receiver,
         uint256 swapAmount,
         bool returnTokenIsCollateral,
         bytes memory loanDataBytes
-    )
-        internal
-        returns (
-            uint256 loanCloseAmount,
-            uint256 withdrawAmount,
-            address withdrawToken
-        )
-    {
+    ) internal returns (CloseWithSwapReturn memory closeWithSwapReturn) {
         require(swapAmount != 0, "swapAmount == 0");
 
         (Loan storage loanLocal, LoanParams storage loanParamsLocal) = _checkLoan(loanId);
 
+        /// Cache the Loan Token Price
+        address loanTokenPoolAddress = underlyingToLoanPool[loanParamsLocal.loanToken];
+        require(loanTokenPoolAddress != address(0), "Invalid loan token pool address");
+
+        TempLoanPoolBalance memory loanPoolBalance =
+            TempLoanPoolBalance({
+                initialLoanTokenSupply: ILoanTokenModules(loanTokenPoolAddress).totalSupply(),
+                latestLoanTokenSupply: 0
+            });
+
```

Comment:
could you move the invariant checks to a modifier, please?

````
modifier iTokenSupplyUnchanged(address loanId){
    check initial balance
    _;
    ensure that final balance == initial balance
}
````
this would make the whole thing much easier readable and reusable.

Reply:
Moving it to modifier might cost another gas, since we need to read the loan & loanParams from the storage, which kinda redundant since we do it inside the function as well.
Wdyt?

Reply:
Updated to modifier